### PR TITLE
fix(types): add TypeScript 4.8 compatibility

### LINKED ISSUE #14934: Typescript 4.8.2 does not like the type definitions from sequelize 6.21.4
Upgrading to Typescripte 4.8.2 and running tsc prints on my system:
```
node_modules/sequelize/types/hooks.d.ts:69:19 - error TS2344: Type 'TAttributes' does not satisfy the constraint '{}'.                  
                                                                                                                                        
69   M extends Model<TAttributes, TCreationAttributes> = Model,                                                                         
                     ~~~~~~~~~~~                                                                                                        
                                                                                                                                        
  node_modules/sequelize/types/hooks.d.ts:70:3                                                                                          
    70   TAttributes = any,                                                                                                             
         ~~~~~~~~~~~~~~~~~                                                                                                              
    This type parameter might need an `extends {}` constraint.                                                                          
                                                                                                                                        
node_modules/sequelize/types/model.d.ts:3252:105 - error TS2344: Type 'TModelAttributes' does not satisfy the constraint '{}'.          
                                                                                                                                        
3252 export type ModelType<TModelAttributes = any, TCreationAttributes = TModelAttributes> = new () => Model<TModelAttributes, TCreation
Attributes>;
                                                                                                             ~~~~~~~~~~~~~~~~

  node_modules/sequelize/types/model.d.ts:3252:23
    3252 export type ModelType<TModelAttributes = any, TCreationAttributes = TModelAttributes> = new () => Model<TModelAttributes, TCrea
tionAttributes>;
                               ~~~~~~~~~~~~~~~~~~~~~~
    This type parameter might need an `extends {}` constraint.

node_modules/sequelize/types/model.d.ts:3260:52 - error TS2344: Type 'S' does not satisfy the constraint '{}'.

3260 export type ModelDefined<S, T> = ModelStatic<Model<S, T>>;
                                                        ~

  node_modules/sequelize/types/model.d.ts:3260:26
    3260 export type ModelDefined<S, T> = ModelStatic<Model<S, T>>;
                                  ~
    This type parameter might need an `extends {}` constraint.
```
As these errors do not reference my files, I assume that the files with type definitions exported by sequelize 6.21.4 are not good.

### PR DESCRIPTION
**Pull request by bnorbi95**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [ ] Have you added new tests to prevent regressions?
- [ ] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Sequelize typings are affected in a few places by a TypeScript 4.8 breaking change (https://devblogs.microsoft.com/typescript/announcing-typescript-4-8/#unconstrained-generics-no-longer-assignable-to).

The PR brings changes from https://github.com/sequelize/sequelize/pull/14990 (which closes #14934) to `main` branch.

### CODE REVIEW

**Comment by ephys** [comment]
Thank you!

Can you remove TS 4.4 from this array, and add 4.8 at its end?
https://github.com/sequelize/sequelize/blob/main/.github/workflows/ci.yml#L116

**Review by ephys — APPROVED** [review]

File: yarn.lock [hunk 1]
```diff
@@ -6210,7 +6210,6 @@ npm@^8.3.0:
     "@npmcli/fs" "^2.1.0"
     "@npmcli/map-workspaces" "^2.0.3"
     "@npmcli/package-json" "^2.0.0"
```
Comment (WikiRik) [inline yarn.lock:6212]: Can we undo these changes? They're not related to the TS update

**Comment by WikiRik** [comment]
Also, I think it would be better if this is a `fix`. But we can update the PR title before merging

**Comment by ephys** [comment]
> Also, I think it would be better if this is a fix

Damn I should have read this before merging the other one, v6 was merged as feat

File: yarn.lock (hunk 1 above)
Comment (bnorbi95) [reply yarn.lock:6212]: Updating TypeScript actually makes these changes to `yarn.lock`. Do you still want me to undo?

**Comment by bnorbi95** [comment]
I've updated commit message and PR title to use `fix`.

**Review by ephys — APPROVED** [review]

File: yarn.lock (hunk 1 above)
Comment (WikiRik) [reply yarn.lock:6212]: Interesting. Can you manually add the three lines back? 6213, 6224 and 6244

Comment (bnorbi95) [reply yarn.lock]: Done!

**Review by WikiRik — APPROVED** [review]

**Review by WikiRik — CHANGES_REQUESTED** [review]
Interesting that it doesn't pass the lint anymore. Can you remove `as any` from the `src/utils/sql.ts` file

**Comment by ephys** [comment]
I think that's going to cause older typescript versions to fail. Muting the lint error could resolve the issue

**Comment by WikiRik** [comment]
That's very possible. Muting would also be fine for now

**Comment by bnorbi95** [comment]
There was an additional error related to https://devblogs.microsoft.com/typescript/announcing-typescript-4-8/#lib-d-ts-updates.

Ready to review again!

**Review by WikiRik — APPROVED** [review]

**Comment by ephys** [comment]
Thank you for making this PR :)
